### ISSUE TITLE
Report document outline

### ISSUE THREAD

**Issue by wangkuiyi** [sql-machine-learning/elasticdl#2022]

**Comment by wangkuiyi** [comment]
ElasticDL: 像写单机程序一样写分布式深度学习程序

- 深度学习的分布式程序 for Kubernetes 不容易写
  - Horovod (AllReduce-only)
  - Distributed Strategy
  - Kubeflow
  - ElasticDL

- ElasticDL 是如何解决问题的
  - worker call TF eager
  - master provides the train loop

- Parameter server 的改进

**Comment by wangkuiyi** [comment]
ElasticDL：同时提升研发效率和机群利用率

**Comment by QiJune** [comment]
### Horovod

只支持allreduce

#### 对TensorFlow 2.0的支持

一个[例子](https://horovod.readthedocs.io/en/stable/tensorflow.html)

```python
import tensorflow as tf
import horovod.tensorflow as hvd

# Initialize Horovod
hvd.init()

# Pin GPU to be used to process local rank (one GPU per process)
gpus = tf.config.experimental.list_physical_devices('GPU')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)
if gpus:
    tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], 'GPU')

# Build model and dataset
dataset = ...
model = ...
loss = tf.losses.SparseCategoricalCrossentropy()
opt = tf.optimizers.Adam(0.001 * hvd.size())

checkpoint_dir = './checkpoints'
checkpoint = tf.train.Checkpoint(model=model, optimizer=opt)

@tf.function
def training_step(images, labels, first_batch):
    with tf.GradientTape() as tape:
        probs = mnist_model(images, training=True)
        loss_value = loss(labels, probs)

    # Horovod: add Horovod Distributed GradientTape.
    tape = hvd.DistributedGradientTape(tape)

    grads = tape.gradient(loss_value, mnist_model.trainable_variables)
    opt.apply_gradients(zip(grads, mnist_model.trainable_variables))

    # Horovod: broadcast initial variable states from rank 0 to all other processes.
    # This is necessary to ensure consistent initialization of all workers when
    # training is started with random weights or restored from a checkpoint.
    #
    # Note: broadcast should be done after the first gradient step to ensure optimizer
    # initialization.
    if first_batch:
        hvd.broadcast_variables(mnist_model.variables, root_rank=0)
        hvd.broadcast_variables(opt.variables(), root_rank=0)

    return loss_value

# Horovod: adjust number of steps based on number of GPUs.
for batch, (images, labels) in enumerate(dataset.take(10000 // hvd.size())):
    loss_value = training_step(images, labels, batch == 0)

    if batch % 10 == 0 and hvd.local_rank() == 0:
        print('Step #%d\tLoss: %.6f' % (batch, loss_value))

# Horovod: save checkpoints only on worker 0 to prevent other workers from
# corrupting it.
if hvd.rank() == 0:
    checkpoint.save(checkpoint_dir)
```

#### 如何启动分布式作业

- horovodrun命令行工具
- [MPI operator](https://github.com/kubeflow/mpi-operator/) of Kubeflow

**Comment by QiJune** [comment]
### TensorFlow distribute strategy

[distribute strategy类型](https://www.tensorflow.org/guide/distributed_training#types_of_strategies)

|  Training API   | MirroredStrategy  | MultiWorkerMirroredStrategy| ParameterServerStrategy|
|  ----  | ----  | --- | --- | 
| Keras API | Supported |  Experimental support| Supported planned post 2.3|
| Custom training loop | Supported |Experimental support | Supported planned post 2.3|
| Estimator API | Limited support| Limited support|  Limited support| 


MirroredStrategy是单机多卡的allreduce，MultiWorkerMirroredStrategy是多机多卡的allreduce。

一个在Kubernetes上跑MirroredStrategy的end to end的例子 https://github.com/tensorflow/ecosystem/tree/master/distribution_strategy

**Comment by QiJune** [comment]
### Kubeflow

#### TensorFlow 1.x

可以支持Parameter Server方式的分布式训练，一个例子https://github.com/kubeflow/tf-operator/tree/master/examples/v1/dist-mnist


#### TensorFlow 2.0

有MirroredStrategy/MultiWorkerMirroredStrategy的例子，https://github.com/kubeflow/tf-operator/tree/master/examples/v1/distribution_strategy

**Comment by workingloong** [comment]
TensorFlow Parameter Server 分布式训练。

TensorFlow 的官网介绍，Keras API 暂时还不支持 Parameter Server 分布式训练。

| Training API | MirroredStrategy | TPUStrategy | MultiWorkerMirroredStrategy | CentralStorageStrategy | ParameterServerStrategy | OneDeviceStrategy |
|--------|--------|-------|--------|-------|---------|-------- |
| **Keras API**  | Supported |  Supported | Experimental support | Experimental support | Supported planned post 2.3 | Supported |
| **Custom training loop** | Supported | Supported | Experimental support    | Experimental support  | Supported planned post 2.3 | Supported |
| **Estimator API** | Limited Support | Not supported | Limited Support | Limited Support | Limited Support | Limited Support |

使用 Keras API 定义的模型如果需要使用 ParameterServerStrategy 分布式训练，需要将
模型转成 estimator。Kubeflow 提供了使用 estimator 来使用 ParameterServerStrategy 分布式训练的例子。
https://github.com/kubeflow/tf-operator/blob/master/examples/v1/distribution_strategy/estimator-API/keras_model_to_estimator.py

```python
def main(args):
  # Define DistributionStrategies and convert the Keras Model to an
  # Estimator that utilizes these DistributionStrateges.
  # Evaluator is a single worker, so using MirroredStrategy.
  config = tf.estimator.RunConfig(
      experimental_distribute=tf.contrib.distribute.DistributeConfig(
          train_distribute=tf.contrib.distribute.CollectiveAllReduceStrategy(
              num_gpus_per_worker=0),
          eval_distribute=tf.contrib.distribute.MirroredStrategy(
              num_gpus_per_worker=0)))
  keras_estimator = tf.keras.estimator.model_to_estimator(
      keras_model=model, config=config, model_dir=model_dir)

  # Train and evaluate the model. Evaluation will be skipped if there is not an
  # "evaluator" job in the cluster.
  tf.estimator.train_and_evaluate(
      keras_estimator,
      train_spec=tf.estimator.TrainSpec(input_fn=input_fn),
      eval_spec=tf.estimator.EvalSpec(input_fn=input_fn))

```

用户除了 keras 模型定义还需要做如下配置：
1. 配置 tf.estimator.RunConfig()
2. 将 keras model 转成 estimator
3. 执行 tf.estimator_train_and_evalute()

Estimator API 仅支持 graph execution，不支持 eager execution，调试代码和网络各层输出比较麻烦。

**Comment by skydoorkai** [comment]
Horovod supports both TensorFlow 1.x and 2.x.

Users can extend a TensorFlow non-distributed training into allreduce-based distributed training using Horovod by two steps:

1. In the first batch, make all workers have the same model by a broadcast.

1.  In training, every worker will get allreduced gradients before apply gradients with a TensorFlow optimizer. 
 
- In TensorFlow 1.x, this is done by wrapping an optimizer:

```
opt = hvd.DistributedOptimizer(opt)
```

- In TensorFlow 2.x, this is done by wrapping the tape:

```
tape = hvd.DistributedGradientTape(tape)
```

**Comment by workingloong** [comment]
> ElasticDL：同时提升研发效率和集群利用率

- 背景介绍。Kubernetes 是目前最先进的分布式操作系统，是公有云和私有云的事实工业标准。目前在 Kubernetes 上运行 TensorFlow 作业的框架有 Kubeflow 的 TF-operator， 但是其属于刚性调度(gang scheduling)，很难同时兼顾研发效率和集群利用率。如果是共享集群资源，资源不足导致长时间等待，影响研发效率。如果是专有资源池，作业拉起快，但是资源利用率低。ElasticDL 利用 TensorFlow eager execution 和 Kubernetes API，可以在集群资源不足时拉起训练作业，当有其他作业释放资源时，能自动扩容到所申请的资源量，这样既能缩短用户作业等待时间，也能提升集群资源利用率。

- ElasticDL 弹性调度的设计。ElasticDL 通过在 Kubernetes 上创建 master 进程来控制作业的弹性调度。这个 master 实现动态数据分片和分发机制，可以把失败进程的训练任务能重新分配给其他进程。master 也能通过 Kubernetes API 获取集群状态，当集群资源充沛时，可以扩容训练进程。

- ElasticDL 弹性调度的 benchmark 实验，证明 ElasticDL 的弹性调度能达到提供研发效率和集群资源利用率的目的

- ElasticDL 在蚂蚁的落地案列，详细介绍花呗的 DBAN 模型。

待讨论：
背景介绍中是否需要讨论 TorchElastic？ TorchElastic 在启动分布式作业时，需要指定进程数变化的范围，任何一个进程均可加入或者脱离作业。当进程数据量发生变化时，任务中的进程需要重新协商，建立新的作业，然后从之前的某个 checkpoint 恢复模型。在 Kubernetes 优先级抢占环境下，进程数量可能变化频繁，其需要反复停止训练并重新开始，弹性效率较低；同时如果所剩进程数少于下限，训练作业也将会失败。

TorchElastic 也是弹性调度的一种实现机制，实现方案与 ElasticDL  不同，其支持的是 AllReduce 策略，没有提到如何支持 ParameterServer。如果我们的文章是给 TensorFlow 社区的，感觉可以不用在文章里提 TorchElastic。

**Comment by wangkuiyi** [comment]
> 背景介绍中是否需要讨论 TorchElastic？

如果集中在 TF+ Kubernetes 的语境下，确实不需要讨论 TorchElastic。而且我们对这个系统了解可能也不够深刻，不宜评价。此外，如果文章发表在  Google 的媒体上，评价 Facebook 的技术，也不是一个合理的做法。
